soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -166,6 +166,37 @@ bool intel_adsp_ipc_send_message_sync(const struct device *dev,
 	return ret;
 }
 
+void intel_adsp_ipc_send_message_emergency(const struct device *dev, uint32_t data,
+					   uint32_t ext_data)
+{
+	const struct intel_adsp_ipc_config * const config = dev->config;
+
+	volatile struct intel_adsp_ipc * const regs = config->regs;
+	bool done;
+
+	/* check if host is processing message. */
+	while (regs->idr & INTEL_ADSP_IPC_BUSY) {
+		k_busy_wait(1);
+	}
+
+	/* check if host has pending acknowledge msg
+	 * Same signal, but on different bits in 1.5
+	 */
+	done = IS_ENABLED(CONFIG_SOC_INTEL_CAVS_V15) ? (regs->idd & INTEL_ADSP_IPC_DONE)
+						     : (regs->ida & INTEL_ADSP_IPC_DONE);
+	if (done) {
+		/* IPC completion */
+		if (IS_ENABLED(CONFIG_SOC_INTEL_CAVS_V15)) {
+			regs->idd = INTEL_ADSP_IPC_DONE;
+		} else {
+			regs->ida = INTEL_ADSP_IPC_DONE;
+		}
+	}
+
+	regs->idd = ext_data;
```

Comment:
is it right ?  this is overwriting the value set few lines above  for `regs->idd` (CAVS_V15 build)

Reply:
Good question. I made it identical to the `intel_adsp_ipc_send_message` function, which uses these registers regardless of the platform. I suspect that setting this bit triggers message acknowledgment, and it doesn't need to stay set. Maybe @aborisovich can confirm this?

Reply:
Yes it looks like by analogy it should work, I'll look into TGL registers asap.

Reply:
On CAVS1.5, IDD register is also the DIPCT register that has a 31 bit as busy bit. IPC Initiator (host) is sets this bit when intents to send the message to the DSP core and we clear this bit in fw when message is accepted.
So the current solution is ok since we just check in line 185 whether there is no incoming messages from ipc initiator for cavs1.5.

Reply:
Ok, so we have it wrong.
IDD register on CAVS1.5 is also DIPCIE register.
DIPCIE register has bit 30 used as done bit. When set, it means that ipc target (host here) had completed the operation. DIPCIE message extension is composed of 29 youngest bits.
So basically in this code, we override DIPCIE bits 30 and 31 using extended message what should not happen.
ext_data should be truncated from the left  to contain only 29 youngest bits of the extended message for CAVS1.5.